analyzers/src/SonarAnalyzer.CFG/ShimLayer/AnalysisContext/SymbolStartAnalysisContext.cs

```diff
@@ -0,0 +1,169 @@
+        registerCodeBlockAction = CreateRegistrationMethod<CodeBlockAnalysisContext>(symbolStartAnalysisContextType, nameof(RegisterCodeBlockAction));
+        registerCodeBlockStartActionCS =
+            CreateRegistrationMethod<CodeBlockStartAnalysisContext<CS.SyntaxKind>>(symbolStartAnalysisContextType, nameof(RegisterCodeBlockStartAction), typeof(CS.SyntaxKind));
+        registerOperationAction =
+            CreateRegistrationMethodWithAdditionalParameter<OperationAnalysisContext, ImmutableArray<OperationKind>>(symbolStartAnalysisContextType, nameof(RegisterOperationAction));
+        registerOperationBlockAction = CreateRegistrationMethod<OperationBlockAnalysisContext>(symbolStartAnalysisContextType, nameof(RegisterOperationBlockAction));
+        registerOperationBlockStartAction = CreateRegistrationMethod<OperationBlockStartAnalysisContext>(symbolStartAnalysisContextType, nameof(RegisterOperationBlockStartAction));
+        registerSymbolEndAction = CreateRegistrationMethod<SymbolAnalysisContext>(symbolStartAnalysisContextType, nameof(RegisterSymbolEndAction));
+        registerSyntaxNodeActionCS = CreateRegistrationMethodWithAdditionalParameter<SyntaxNodeAnalysisContext, ImmutableArray<CS.SyntaxKind>>(
+            symbolStartAnalysisContextType, nameof(RegisterSyntaxNodeAction), typeof(CS.SyntaxKind));
+    }
+
+    public SymbolStartAnalysisContext(object roslynSymbolStartAnalysisContext)
+    {
+        RoslynSymbolStartAnalysisContext = roslynSymbolStartAnalysisContext;
+    }
+    public object RoslynSymbolStartAnalysisContext { get; }
+    public CancellationToken CancellationToken => cancellationTokenAccessor(RoslynSymbolStartAnalysisContext);
+    public Compilation Compilation => compilationAccessor(RoslynSymbolStartAnalysisContext);
+    public AnalyzerOptions Options => optionsAccessor(RoslynSymbolStartAnalysisContext);
+    public ISymbol Symbol => symbolAccessor(RoslynSymbolStartAnalysisContext);
+
+    public void RegisterCodeBlockAction(Action<CodeBlockAnalysisContext> action) =>
+        registerCodeBlockAction(RoslynSymbolStartAnalysisContext, action);
+
+    public void RegisterCodeBlockStartAction<TLanguageKindEnum>(Action<CodeBlockStartAnalysisContext<TLanguageKindEnum>> action) where TLanguageKindEnum : struct
+    {
+        var languageKindType = typeof(TLanguageKindEnum);
+        if (languageKindType == typeof(CS.SyntaxKind))
+        {
+            var casted = (Action<CodeBlockStartAnalysisContext<CS.SyntaxKind>>)action;
+            registerCodeBlockStartActionCS(RoslynSymbolStartAnalysisContext, casted);
+        }
+        else if (languageKindType.FullName == "Microsoft.CodeAnalysis.VisualBasic.SyntaxKind")
+        {
+            throw new NotImplementedException("Add a reference to the Microsoft.CodeAnalysis.VisualBasic.Workspaces package.");
+        }
+        else
+        {
+            throw new ArgumentException("Invalid type parameter.", nameof(TLanguageKindEnum));
```

Comment:
On this line we have "[S3928](https://rules.sonarsource.com/csharp/RSPEC-3928/) Parameter names used into ArgumentException constructors should match an existing one" triggering on `nameof(TLanguageKindEnum)`, because `TLanguageKindEnum` is a type argument, not a value argument.

I wonder if that's an acceptable use of `ArgumentException` (that is normally used to validate value arguments, not type arguments). 

I see that you have expressed your opinion by creating [this issue](https://github.com/SonarSource/sonar-dotnet/issues/8822).
I think I would agree.